hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/CompletableFutureUsageCondition.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.test.archunit;
+
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaMember;
+import com.tngtech.archunit.core.domain.JavaMethodCall;
+import com.tngtech.archunit.core.domain.JavaType;
+import com.tngtech.archunit.core.importer.ClassFileImporter;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ConditionEvents;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.tngtech.archunit.lang.SimpleConditionEvent.violated;
+
+/**
+ * <ul>
+ * <li> from {@link CompletionStage} create a list of methods that have an {@code Async} counterpart
+ * <li> based on this list, filtering all their calls on the {@link CompletableFuture} {@code instanceof} objects
+ * <li> checking that no non-Async methods versions are used
+ * <li> checking that for Async methods the {@link Executor} service is specified
+ * <li> skipping methods that override {@link CompletableFuture} base class methods
```

Comment:
Why `skipping methods that override {@link CompletableFuture} base class methods` ?

Reply:
Because in our classes that extend the `CompletableFuture`, we override the base class methods, and those cases shouldn’t be reported by ArchUnit.

For example like this one: https://github.com/hazelcast/hazelcast/blob/7c3894b1474dba56648cf10ad8441615486fcdea/hazelcast/src/main/java/com/hazelcast/spi/impl/DelegatingCompletableFuture.java#L235-L238

Reply:
So maybe exclude those 2 classes? `DelegatingCompletableFuture` and `InternalCompletableFuture`

Reply:
In addition to `DelegatingCompletableFuture` and `InternalCompletableFuture`, there are ~15 classes that extend the `InternalCompletableFuture`.  
Inside those classes, except methods that override the `CompletableFuture` base methods, also present methods where we should also replace the sync/async methods invocation.